pkgs/build-support/node/fetch-yarn-deps/yarn-install-hook.sh

```diff
@@ -0,0 +1,82 @@
+# shellcheck shell=bash
+
+yarnInstallHook() {
+    echo "Executing yarnInstallHook"
+
+    runHook preInstall
+
+    local -r packageOut="$out/lib/node_modules/$(@jq@ --raw-output '.name' package.json)"
+
+    # `npm pack` writes to cache so temporarily override it
+    while IFS= read -r file; do
+        local dest="$packageOut/$(dirname "$file")"
+        mkdir -p "$dest"
+        cp "$file" "$dest"
+    done < <(@jq@ --raw-output '.[0].files | map(.path | select(. | startswith("node_modules/") | not)) | join("\n")' <<< "$(npm_config_cache="$HOME/.npm" npm pack --json --dry-run --loglevel=warn --no-foreground-scripts $npmPackFlags "${npmPackFlagsArray[@]}" $npmFlags "${npmFlagsArray[@]}")")
+
+    # Based on code from Python's buildPythonPackage wrap.sh script, for
+    # supporting both the case when makeWrapperArgs is an array and a
+    # IFS-separated string.
+    #
+    # TODO: remove the string branch when __structuredAttrs are used.
+    if [[ "${makeWrapperArgs+defined}" == "defined" && "$(declare -p makeWrapperArgs)" =~ ^'declare -a makeWrapperArgs=' ]]; then
+        local -a user_args=("${makeWrapperArgs[@]}")
+    else
+        local -a user_args="(${makeWrapperArgs:-})"
+    fi
+    while IFS=" " read -ra bin; do
+        mkdir -p "$out/bin"
+        makeWrapper @hostNode@ "$out/bin/${bin[0]}" --add-flags "$packageOut/${bin[1]}" "${user_args[@]}"
+    done < <(@jq@ --raw-output '(.bin | type) as $typ | if $typ == "string" then
+        .name + " " + .bin
+        elif $typ == "object" then .bin | to_entries | map(.key + " " + .value) | join("\n")
+        elif $typ == "null" then empty
+        else "invalid type " + $typ | halt_error end' "package.json")
+
+    while IFS= read -r man; do
+        installManPage "$packageOut/$man"
+    done < <(@jq@ --raw-output '(.man | type) as $typ | if $typ == "string" then .man
+        elif $typ == "list" then .man | join("\n")
+        elif $typ == "null" then empty
+        else "invalid type " + $typ | halt_error end' "package.json")
+
+    local -r nodeModulesPath="$packageOut/node_modules"
+
+    if [ ! -d "$nodeModulesPath" ]; then
+        if [ -z "${dontYarnPrune-}" ]; then
+            # Yarn has a 'prune' command, but it's only a stub that directs you to use install
```

Comment:
Could you explain these conditionals?

1. Wouldn't the commands below fail more informatively if the node_modules directory wouldn't exist and you wouldn't silently don't do anything in that case?
2. `dontYarnPrune` is not documented, and after reading the comment it seems confusing to me - we have `dontYarnInstall` already which achieves the same behavior.

Reply:
> Could you explain these conditionals?
> 
>     1. Wouldn't the commands below fail more informatively if the node_modules directory wouldn't exist and you wouldn't silently don't do anything in that case?

This is a vestige from npmInstallHook, `npm prune` fails if `node_modules` does not exist AFIK. 
Still, I think it makes sense to keep this as is, `node_modules` should be created by `yarnConfigHook`, in the edge case that `yarnInstallHook` does not see one, I think it would be unexpected for it to create a "new" `node_modules` by itself. 
If there was a `yarn prune` command, I would be using that, and then, this check would be mandatory, is what I'm trying to say.

>     2. `dontYarnPrune` is not documented, and after reading the comment it seems confusing to me - we have `dontYarnInstall` already which achieves the same behavior.

The documentation is [here](https://github.com/NixOS/nixpkgs/pull/328544/files#diff-69c4ba235ab61f76ad80555351e0567178235fce8b8f966ff689e758e1beb771R445), I added it later.

We need both `dontYarnInstall` and `dontYarnPrune` to support packages that incorrectly put runtime dependencies in `devDependencies`. Ideally we wouldn't need `dontYarnPrune`, but I'm not sure we can remove it, so I'm playing it safe.